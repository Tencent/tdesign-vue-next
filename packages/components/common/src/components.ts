// 基础
export * from './button';
export * from './icon';
export * from './link';

// 布局
export * from './grid';
export * from './layout';
export * from './divider';
export * from './space';
export * from './descriptions';

// 导航
export * from './affix';
export * from './anchor';
export * from './breadcrumb';
export * from './dropdown';
export * from './menu';
export * from './pagination';
export * from './steps';
export * from './tabs';
export * from './sticky-tool';
export * from './back-top';

// 输入
export * from './auto-complete';
export * from './cascader';
export * from './checkbox';
export * from './date-picker';
export * from './form';
export * from './input-adornment';
export * from './input';
export * from './input-number';
export * from './radio';
export * from './select';
export * from './slider';
export * from './switch';
export * from './tag-input';
export * from './select-input';
export * from './textarea';
export * from './transfer';
export * from './time-picker';
export * from './tree-select';
export * from './color-picker';
export * from './range-input';

// 数据展示
export * from './avatar';
export * from './badge';
export * from './calendar';
export * from './card';
export * from './comment';
export * from './image';
export * from './image-viewer';
export * from './list';
export * from './progress';
export * from './skeleton';
export * from './statistic';
export * from './swiper';
export * from './table';
export * from './tag';
export * from './timeline';
export * from './tooltip';
export * from './tree';
export * from './collapse';
export * from './watermark';
export * from './rate';

// 消息提醒

export * from './alert';
export * from './dialog';
export * from './drawer';
export * from './guide';
export * from './loading';
export * from './message';
export * from './notification';
export * from './popconfirm';
export * from './popup';
export * from './upload';

// 全局配置
export * from './config-provider';
