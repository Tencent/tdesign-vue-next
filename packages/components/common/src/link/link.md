:: BASE_DOC ::

## API
### Link Props

名称 | 类型 | 默认值 | 说明 | 必传
-- | -- | -- | -- | --
content | String / Slot / Function | - | 链接内容。TS 类型：`string \| TNode`。[通用类型定义](https://github.com/Tencent/tdesign-vue-next/blob/develop/src/common.ts) | N
default | String / Slot / Function | - | 链接内容，同 content。TS 类型：`string \| TNode`。[通用类型定义](https://github.com/Tencent/tdesign-vue-next/blob/develop/src/common.ts) | N
disabled | Boolean | undefined | 禁用链接。优先级：Link.disabled > Form.disabled | N
download | String / Boolean | - | 使得浏览器将链接的 URL 视为可下载资源 | N
hover | String | underline | 链接悬浮态样式，有 文本颜色变化、添加下划线等 2 种方法。可选项：color/underline | N
href | String | - | 跳转链接 | N
prefixIcon | Slot / Function | - | 前置图标。TS 类型：`TNode`。[通用类型定义](https://github.com/Tencent/tdesign-vue-next/blob/develop/src/common.ts) | N
size | String | medium | 尺寸。可选项：small/medium/large。TS 类型：`SizeEnum`。[通用类型定义](https://github.com/Tencent/tdesign-vue-next/blob/develop/src/common.ts) | N
suffixIcon | Slot / Function | - | 后置图标。TS 类型：`TNode`。[通用类型定义](https://github.com/Tencent/tdesign-vue-next/blob/develop/src/common.ts) | N
target | String | - | 跳转方式，如：当前页面打开、新页面打开等，同 HTML 属性 target 含义相同 | N
theme | String | default | 组件风格，依次为默认色、品牌色、危险色、警告色、成功色。可选项：default/primary/danger/warning/success | N
underline | Boolean | - | 普通状态是否显示链接下划线 | N
onClick | Function |  | TS 类型：`(e: MouseEvent) => void`<br/>点击事件，禁用状态不会触发点击事件 | N

### Link Events

名称 | 参数 | 描述
-- | -- | --
click | `(e: MouseEvent)` | 点击事件，禁用状态不会触发点击事件
