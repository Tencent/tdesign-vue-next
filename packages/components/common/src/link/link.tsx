import { computed, defineComponent } from '@td/adapter-vue';
import { useCommonClassName, useConfig, useContent, useDisabled, usePrefixClass, useTNodeJSX } from '@td/adapter-hooks';
import props from '@td/intel/link/props';

export default defineComponent({
  name: 'TLink',
  props: { ...props },
  emits: ['click'],
  setup(props, { emit }) {
    const renderContent = useContent();
    const renderTNodeJSX = useTNodeJSX();
    const COMPONENT_NAME = usePrefixClass('link');
    const { STATUS, SIZE } = useCommonClassName();
    const { classPrefix } = useConfig('classPrefix');
    const isDisabled = useDisabled();

    const linkClass = computed(() => [
      `${COMPONENT_NAME.value}`,
      `${COMPONENT_NAME.value}--theme-${props.theme}`,
      {
        [SIZE.value[props.size]]: props.size !== 'medium',
        [STATUS.value.disabled]: isDisabled.value,
        [`${classPrefix.value}-is-underline`]: props.underline,
        [`${COMPONENT_NAME.value}--hover-${props.hover}`]: !isDisabled.value,
      },
    ]);
    // 禁用时 无点击事件
    const handleClick = (event: MouseEvent) => {
      if (!isDisabled.value) {
        emit('click', event);
      }
    };
    return () => {
      const linkContent = renderContent('default', 'content');
      const prefix = renderTNodeJSX('prefixIcon');
      const suffix = renderTNodeJSX('suffixIcon');

      return (
        <a
          class={[...linkClass.value]}
          href={isDisabled.value || !props.href ? undefined : props.href}
          target={!props.target ? undefined : props.target}
          download={!props.download ? undefined : props.download}
          onClick={handleClick}
        >
          {prefix ? <span class={`${COMPONENT_NAME.value}__prefix-icon`}>{prefix}</span> : null}
          {linkContent}
          {suffix ? <span class={`${COMPONENT_NAME.value}__suffix-icon`}>{suffix}</span> : null}
        </a>
      );
    };
  },
});
