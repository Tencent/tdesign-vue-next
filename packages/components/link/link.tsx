import { computed, defineComponent } from 'vue';
import { useConfig, usePrefixClass, useCommonClassName } from '../hooks/useConfig';
import { useContent, useTNodeJSX } from '../hooks/tnode';
import { useDisabled } from '../hooks/useDisabled';
import props from './props';

export default defineComponent({
  name: 'TLink',
  props: { ...props },
  emits: ['click'],
  setup(props, { emit }) {
    const renderContent = useContent();
    const renderTNodeJSX = useTNodeJSX();
    const COMPONENT_NAME = usePrefixClass('link');
    const { STATUS, SIZE } = useCommonClassName();
    const { classPrefix } = useConfig('classPrefix');
    const isDisabled = useDisabled();

    const linkClass = computed(() => [
      `${COMPONENT_NAME.value}`,
      `${COMPONENT_NAME.value}--theme-${props.theme}`,
      {
        [SIZE.value[props.size]]: props.size !== 'medium',
        [STATUS.value.disabled]: isDisabled.value,
        [`${classPrefix.value}-is-underline`]: props.underline,
        [`${COMPONENT_NAME.value}--hover-${props.hover}`]: !isDisabled.value,
      },
    ]);
    // 禁用时 无点击事件
    const handleClick = (event: MouseEvent) => {
      if (!isDisabled.value) emit('click', event);
    };
    return () => {
      const linkContent = renderContent('default', 'content');
      const prefix = renderTNodeJSX('prefixIcon');
      const suffix = renderTNodeJSX('suffixIcon');

      return (
        <a
          class={[...linkClass.value]}
          href={isDisabled.value || !props.href ? undefined : props.href}
          target={!props.target ? undefined : props.target}
          download={!props.download ? undefined : props.download}
          onClick={handleClick}
        >
          {prefix ? <span class={`${COMPONENT_NAME.value}__prefix-icon`}>{prefix}</span> : null}
          {linkContent}
          {suffix ? <span class={`${COMPONENT_NAME.value}__suffix-icon`}>{suffix}</span> : null}
        </a>
      );
    };
  },
});
