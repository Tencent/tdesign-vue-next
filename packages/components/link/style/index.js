import '../../../common/style/web/components/link/_index.less';
