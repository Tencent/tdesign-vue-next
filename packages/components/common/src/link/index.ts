import { withInstall } from '@td/adapter-utils';
import type { TdLinkProps } from '@td/intel/link/type';
import _Link from './link';

import './style';

export * from '@td/intel/link/type';
export type LinkProps = TdLinkProps;

export const Link = withInstall(_Link);
export default Link;
