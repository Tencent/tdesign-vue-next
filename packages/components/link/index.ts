import _Link from './link';
import { withInstall } from '@tdesign/shared-utils';
import { TdLinkProps } from './type';

import './style';

export * from './type';
export type LinkProps = TdLinkProps;

export const Link = withInstall(_Link);
export default Link;
