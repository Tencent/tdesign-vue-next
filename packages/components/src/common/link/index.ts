import _Link from './link';
import { withInstall } from '@td/adapter-utils';
import type { TdLinkProps } from '@td/intel/components/link/type';

import './style';

export * from '@td/intel/components/link/type';
export type LinkProps = TdLinkProps;

export const Link = withInstall(_Link);
export default Link;
