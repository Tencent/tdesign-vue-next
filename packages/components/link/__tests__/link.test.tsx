import { mount } from '@vue/test-utils';
import { vi } from 'vitest';
import { Link, LinkProps } from '@tdesign/components/link';

describe('Link', () => {
  describe('props', () => {
    it('content[function]', () => {
      const wrapper = mount(<Link content={() => 'content test'}></Link>);
      expect(wrapper.find('.t-link').text()).toBe('content test');
    });

    it('content[slot]', () => {
      const wrapper = mount(<Link v-slots={{ content: () => 'content test' }}></Link>);
      expect(wrapper.find('.t-link').text()).toBe('content test');
    });

    it('default[function]', () => {
      const wrapper = mount(<Link default={() => 'default test'}></Link>);
      expect(wrapper.find('.t-link').text()).toBe('default test');
    });

    it('default[slot]', () => {
      const wrapper = mount(<Link v-slots={{ default: () => 'default test' }}></Link>);
      expect(wrapper.find('.t-link').text()).toBe('default test');
    });

    it('disabled[boolean]', () => {
      // disabled default value is
      const wrapper1 = mount(<Link>Text</Link>);
      expect(wrapper1.classes('t-is-disabled')).toBeFalsy();
      // disabled = true
      const wrapper2 = mount(<Link disabled={true}>Text</Link>);
      expect(wrapper2.classes('t-is-disabled')).toBeTruthy();
      expect(wrapper2.element).toMatchSnapshot();
      // disabled = false
      const wrapper3 = mount(<Link disabled={false}>Text</Link>);
      expect(wrapper3.classes('t-is-disabled')).toBeFalsy();
      expect(wrapper3.element).toMatchSnapshot();
    });

    it('hover[string]', () => {
      const hoverList: Array<LinkProps['hover']> = ['color', 'underline'];
      hoverList.forEach((item) => {
        const wrapper = mount(<Link hover={item}>Text</Link>);
        expect(wrapper.classes(`t-link--hover-${item}`)).toBeTruthy();
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    it('href[string]', () => {
      const wrapper = mount(<Link href={'https://tdesign.tencent.com/'}>Text</Link>);
      expect(wrapper.attributes('href')).toBe('https://tdesign.tencent.com/');
      expect(wrapper.element).toMatchSnapshot();
    });

    it('prefixIcon[function]', () => {
      const wrapper = mount(<Link prefixIcon={() => 'prefixIcon'}></Link>);
      expect(wrapper.find('.t-link__prefix-icon').text()).toBe('prefixIcon');
    });

    it('prefixIcon[slot]', () => {
      const wrapper = mount(<Link v-slots={{ prefixIcon: () => 'prefixIcon' }}></Link>);
      expect(wrapper.find('.t-link__prefix-icon').text()).toBe('prefixIcon');
    });

    it('prefix-icon[slot]', () => {
      const wrapper = mount(<Link v-slots={{ 'prefix-icon': () => 'prefixIcon' }}></Link>);
      expect(wrapper.find('.t-link__prefix-icon').text()).toBe('prefixIcon');
    });

    it('size[string]', () => {
      const wrapperSmall = mount(<Link size={'small'}>Text</Link>);
      expect(wrapperSmall.classes('t-size-s')).toBeTruthy();

      const wrapperMedium = mount(<Link size={'medium'}>Text</Link>);
      expect(wrapperMedium.classes('t-size-m')).toBeFalsy();

      const wrapperLarge = mount(<Link size={'large'}>Text</Link>);
      expect(wrapperLarge.classes('t-size-l')).toBeTruthy();
    });

    it('suffixIcon[function]', () => {
      const wrapper = mount(<Link suffixIcon={() => 'suffixIcon'}></Link>);
      expect(wrapper.find('.t-link__suffix-icon').text()).toBe('suffixIcon');
    });

    it('suffixIcon[slot]', () => {
      const wrapper = mount(<Link v-slots={{ suffixIcon: () => 'suffixIcon' }}></Link>);
      expect(wrapper.find('.t-link__suffix-icon').text()).toBe('suffixIcon');
    });
    it('suffix-icon[slot]', () => {
      const wrapper = mount(<Link v-slots={{ 'suffix-icon': () => 'suffix-icon' }}></Link>);
      expect(wrapper.find('.t-link__suffix-icon').text()).toBe('suffix-icon');
    });

    it('target[string]', () => {
      const wrapper = mount(<Link target={'_blank'}>Text</Link>);
      expect(wrapper.attributes('target')).toBe('_blank');
      expect(wrapper.element).toMatchSnapshot();
    });

    it('theme[string]', () => {
      const themeList: Array<LinkProps['theme']> = ['default', 'primary', 'danger', 'warning', 'success'];
      themeList.forEach((item) => {
        const wrapper = mount(<Link theme={item}>Text</Link>);
        expect(wrapper.classes(`t-link--theme-${item}`)).toBeTruthy();
      });
    });

    it('underline[boolean]', () => {
      // underline default value is
      const wrapper1 = mount(<Link>Text</Link>);
      expect(wrapper1.classes('t-is-underline')).toBeFalsy();
      // underline = true
      const wrapper2 = mount(<Link underline={true}>Text</Link>);
      expect(wrapper2.classes('t-is-underline')).toBeTruthy();
      // underline = false
      const wrapper3 = mount(<Link underline={false}>Text</Link>);
      expect(wrapper3.classes('t-is-underline')).toBeFalsy();
    });
  });
  describe('event', () => {
    it('onClick', async () => {
      const fn = vi.fn();
      const wrapper = mount(<Link onClick={fn}></Link>);
      wrapper.findComponent(Link).trigger('click');
      await wrapper.vm.$nextTick();
      expect(fn).toHaveBeenCalled();
      expect(fn.mock.calls[0][0].stopPropagation).toBeTruthy();
      expect(fn.mock.calls[0][0].type).toBe('click');
    });
  });
});
