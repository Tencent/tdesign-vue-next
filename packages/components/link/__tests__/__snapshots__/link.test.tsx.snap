// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > props > disabled[boolean] 1`] = `
<a
  class="t-link t-link--theme-default t-is-disabled"
>
  <!---->
  
  Text
  
  <!---->
</a>
`;

exports[`Link > props > disabled[boolean] 2`] = `
<a
  class="t-link t-link--theme-default t-link--hover-underline"
>
  <!---->
  
  Text
  
  <!---->
</a>
`;

exports[`Link > props > hover[string] 1`] = `
<a
  class="t-link t-link--theme-default t-link--hover-color"
>
  <!---->
  
  Text
  
  <!---->
</a>
`;

exports[`Link > props > hover[string] 2`] = `
<a
  class="t-link t-link--theme-default t-link--hover-underline"
>
  <!---->
  
  Text
  
  <!---->
</a>
`;

exports[`Link > props > href[string] 1`] = `
<a
  class="t-link t-link--theme-default t-link--hover-underline"
  href="https://tdesign.tencent.com/"
>
  <!---->
  
  Text
  
  <!---->
</a>
`;

exports[`Link > props > target[string] 1`] = `
<a
  class="t-link t-link--theme-default t-link--hover-underline"
  target="_blank"
>
  <!---->
  
  Text
  
  <!---->
</a>
`;
